Buttplug.Server/Bluetooth/Devices/Lovense.cs

```diff
@@ -222,10 +239,74 @@ internal class Lovense : ButtplugBluetoothDevice
                 return BpLogger.LogErrorMsg(aMsg.Id, Error.ErrorClass.ERROR_DEVICE, "Wrong Handler");
             }
 
+            return await HandleVibrateCmd(new VibrateCmd(cmdMsg.DeviceIndex,
+                new List<VibrateCmd.VibrateIndex>() { new VibrateCmd.VibrateIndex(0, cmdMsg.Speed) },
+                cmdMsg.Id));
+        }
+
+        private async Task<ButtplugMessage> HandleVibrateCmd(ButtplugDeviceMessage aMsg)
+        {
+            var cmdMsg = aMsg as VibrateCmd;
+            if (cmdMsg is null)
+            {
+                return BpLogger.LogErrorMsg(aMsg.Id, Error.ErrorClass.ERROR_DEVICE, "Wrong Handler");
+            }
+
+            foreach (var vi in cmdMsg.Speeds)
+            {
+                if (vi.Index == 0)
+                {
+                    if (vi.Speed == _vibratorSpeed)
+                    {
+                        return new Ok(cmdMsg.Id);
+                    }
+
+                    _vibratorSpeed = vi.Speed;
+                }
+            }
+
             // While there are 3 lovense revs right now, all of the characteristic arrays are the same.
             return await Interface.WriteValue(aMsg.Id,
                 Info.Characteristics[(uint)LovenseRev1BluetoothInfo.Chrs.Tx],
-                Encoding.ASCII.GetBytes($"Vibrate:{(int)(cmdMsg.Speed * 20)};"));
+                Encoding.ASCII.GetBytes($"Vibrate:{(int)(_vibratorSpeed * 20)};"));
+        }
+
+        private async Task<ButtplugMessage> HandleRotateCmd(ButtplugDeviceMessage aMsg)
+        {
+            var cmdMsg = aMsg as RotateCmd;
+            if (cmdMsg is null)
+            {
+                return BpLogger.LogErrorMsg(aMsg.Id, Error.ErrorClass.ERROR_DEVICE, "Wrong Handler");
+            }
+
+            var dirChange = false;
+            var speedChange = false;
+            foreach (var vi in cmdMsg.Speeds)
+            {
+                if (vi.Index == 0)

```

Comment:
Here and just about everywhere else:

As far as I can tell, there's no updates/errors if someone sends a wrong sub-index for a generic command. It'll just silently not set anything, but it also doesn't return or break or anything, so it'll just send the last set of commands to the toy. This seems bad?